hedera-node/src/main/java/com/hedera/services/store/contracts/HederaWorldState.java

```diff
@@ -417,18 +349,43 @@ public void revert() {
 		@Override
 		public void commit() {
 			final HederaWorldState wrapped = wrappedWorldView();
-			wrapped.provisionalAccountDeletes.addAll(getDeletedAccountAddresses());
 			wrapped.sponsorMap.putAll(sponsorMap);
+			var repository = wrapped.repositoryRoot;
+
+			getDeletedAccountAddresses().forEach(address -> {
+				final var bytesAddress = address.toArray();
+				// account may have been created and deleted within the same transaction.
+				if (repository.isExist(bytesAddress)) {
+					var account = repository.getAccount(bytesAddress);
+					// selfdestruct adjusts balances to zero
+					account.setBalance(BigInteger.ZERO);
+					validateTrue(!repository.isDeleted(bytesAddress), FAIL_INVALID);
+					account.setDeleted(true);
+				}
+			});
 
 			for (final UpdateTrackingAccount<WorldStateAccount> updated : getUpdatedAccounts()) {
-				// Save the code in storage ...
-				if (updated.codeWasUpdated()) {
-					wrapped.provisionalCodeUpdates.put(updated.getAddress(), updated.getCode());
+				final byte[] address = updated.getAddress().toArray();
+
+				AccountState updatedAccount;
+				if (repository.isExist(address)) {
+					updatedAccount = repository.getAccount(address);
+				} else {
+					repository.delete(address);
+					updatedAccount = repository.createAccount(address);
+					wrapped.privisionalContractCreations.add(contractParsedFromSolidityAddress(address));
+				}
+
+				updatedAccount.setBalance(updated.getBalance().toBigInteger());
+				wrapped.ledgerAccountsSource.put(address, updatedAccount);
```

Comment:
I do not think that we need to (or should) use the `ledgerAccountsSource` here and directly work with the `ledger`.  What if we use `repository.addBalance()` method? Is there a problem with it? My thinking is that `ledgerAccountsSource` is an interface that the repository uses and if we can work through the `repository`, it might be better to use that and not skip the caching layer of the repository and go directly to the `ledgerAccountsSource`.

*Yes, there might be no reason for us to put anything in cache only to persist it afterwards on `persist()`. Just sharing an idea